File: algorithms/[15]三数之和/solution.py
# topics = ["数组", "双指针"]

from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        if n < 3:
            return []

        res: List[List[int]] = []
        nums.sort()

        for i in range(n):
            if nums[i] > 0:
                break
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left = i + 1
            right = n - 1
            while left < right:
                sum = nums[i] + nums[left] + nums[right]
                if sum < 0:
                    left += 1
                elif sum > 0:
                    right -= 1
                else:
                    res.append([nums[i], nums[left], nums[right]])
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1
                    left += 1
                    right -= 1

        return res

File: algorithms/[15]三数之和/Solution.cs
// topics = ["数组", "双指针"]

using System;
using System.Collections.Generic;

public class Solution
{
    public IList<IList<int>> ThreeSum(int[] nums)
    {
        IList<IList<int>> res = new List<IList<int>>();

        int n = nums.Length;
        if (n < 3)
        {
            return res;
        }

        Array.Sort(nums);
        for (int i = 0; i < n; i++)
        {
            if (nums[i] > 0) break;
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            int L = i + 1;
            int R = n - 1;

            while (L < R)
            {
                int sum = nums[i] + nums[L] + nums[R];
                if (sum < 0)
                {
                    L++;
                }
                else if (sum > 0)
                {
                    R--;
                }
                else
                {
                    res.Add(new List<int>() { nums[i], nums[L], nums[R] });
                    while (L < R && nums[L] == nums[L + 1]) L++;
                    while (L < R && nums[R] == nums[R - 1]) R--;
                    L++;
                    R--;
                }
            }

        }

        return res;
    }
}
